File: 308__Range_Sum_Query_2D_Mutable.py
class NumMatrix(object):

    def __init__(self, matrix):
        """
        :type matrix: List[List[int]]
        """
        for row in matrix:
            for j in range(1, len(row)):
                row[j] += row[j - 1]
        self.matrix = matrix

    def update(self, row, col, val):
        """
        :type row: int
        :type col: int
        :type val: int
        :rtype: void
        """
        origin = self.matrix[row][col]
        if col != 0:
            origin -= self.matrix[row][col - 1]
        diff = val - origin
        for j in range(col, len(self.matrix[0])):
            self.matrix[row][j] += diff


    def sumRegion(self, row1, col1, row2, col2):
        """
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """
        result = 0
        if col1 != 0:
            for i in range(row1, row2 + 1):
                result += self.matrix[i][col2] - self.matrix[i][col1 - 1]
        else:
            for i in range(row1, row2 + 1):
                result += self.matrix[i][col2]
        return result


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# obj.update(row,col,val)
# param_2 = obj.sumRegion(row1,col1,row2,col2)

File: Java/308__Range_Sum_Query_2D-Mutable.java
class NumMatrix {

    int m, n;
    int[][] data;    // stores matrix[][]
    int[][] BITree; // 2-D binary indexed tree

    public NumMatrix(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0)
            return;
        m = matrix.length;
        n = matrix[0].length;
        data = new int[m][n];
        BITree = new int[m + 1][n + 1];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                update(i, j, matrix[i][j]); // init BITree[][]
                data[i][j] = matrix[i][j];   // init data[][]
            }
        }
    }

    public void update(int i, int j, int val) {
        int diff = val - data[i][j];  // get the diff
        data[i][j] = val;             // update data[][]
        i++; j++;
        while (i <= m) {
            int k = j;
            while (k <= n) {
                BITree[i][k] += diff; // update BITree[][]
                k += k & (-k); // update column index to that of parent
            }
            i += i & (-i);   // update row index to that of parent
        }
    }

    int getSum(int i, int j) {
        int result = 0;
        i++; j++;
        while (i > 0) {
            int k = j;
            while (k > 0) {
                result += BITree[i][k]; // accumulate the sum
                k -= k & (-k); // move column index to parent node
            }
            i -= i & (-i);   // move row index to parent node
        }
        return result;
    }

    public int sumRegion(int i1, int j1, int i2, int j2) {
        return getSum(i2, j2) - getSum(i1 - 1, j2) - getSum(i2, j1 - 1) + getSum(i1 - 1, j1 - 1);
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * obj.update(row,col,val);
 * int param_2 = obj.sumRegion(row1,col1,row2,col2);
 */
